File: JavaScript/0289. Game of Life.js
// According to the Wikipedia's article: "The Game of Life, also known simply as Life, is a cellular automaton devised by the British mathematician John Horton Conway in 1970."
//
// Given a board with m by n cells, each cell has an initial state live (1) or dead (0). Each cell interacts with its eight neighbors (horizontal, vertical, diagonal) using the following four rules (taken from the above Wikipedia article):
//
// Any live cell with fewer than two live neighbors dies, as if caused by under-population.
// 1. Any live cell with two or three live neighbors lives on to the next generation.
// 2. Any live cell with more than three live neighbors dies, as if by over-population..
// 3. Any dead cell with exactly three live neighbors becomes a live cell, as if by reproduction.
// 4. Write a function to compute the next state (after one update) of the board given its current state.
//
// Follow up:
// 1. Could you solve it in-place? Remember that the board needs to be updated at the same time: You cannot update some cells first and then use their updated values to update other cells.
// 2. In this question, we represent the board using a 2D array. In principle, the board is infinite, which would cause problems when the active area encroaches the border of the array. How would you address these problems?

/**
 * @param {number[][]} board
 * @return {void} Do not return anything, modify board in-place instead.
 */
// https://leetcode.com/problems/game-of-life/discuss/73223/Easiest-JAVA-solution-with-explanation
//
// To solve it in place, we use 2 bits to store 2 states:
//
// [2nd bit, 1st bit] = [next state, current state]
//   00  dead (next) <- dead (current)
//   01  dead (next) <- live (current)
//   10  live (next) <- dead (current)
//   11  live (next) <- live (current)
//
// - In the beginning, every cell is either 00 or 01.
// - Notice that 1st state is independent of 2nd state.
// - Imagine all cells are instantly changing from the 1st to the 2nd state, at the same time.
// - Let's count # of neighbors from 1st state and set 2nd state bit.
// - Since every 2nd state is by default dead, no need to consider transition 01 -> 00.
// - In the end, delete every cell's 1st state by doing >> 1.
//
// For each cell's 1st bit, check the 8 pixels around itself, and set the cell's 2nd bit.
//   Transition 01 -> 11: when board == 1 and lives >= 2 && lives <= 3.
//   Transition 00 -> 10: when board == 0 and lives == 3.
//
// To get the current state, simply do
//   board[i][j] & 1
//
// To get the next state, simply do
//   board[i][j] >> 1

const gameOfLife = (board) => {
  if (board == null || board.length === 0) return;

  const h = board.length;
  const w = board[0].length;

  const liveNeighbors = (x, y) => {
    let lives = 0;
    for (let i = Math.max(x - 1, 0); i <= Math.min(x + 1, h - 1); i++) {
      for (let j = Math.max(y - 1, 0); j <= Math.min(y + 1, w - 1); j++) {
        lives += board[i][j] & 1; // Get 1st bit
      }
    }
    lives -= board[x][y] & 1; // subtract itself
    return lives;
  };

  for (let i = 0; i < h; i++) {
    for (let j = 0; j < w; j++) {
      const lives = liveNeighbors(i, j);

      // In the beginning, every 2nd bit is 0
      // So we only need to care about when will the 2nd bit become 1
      if (board[i][j] === 1 && lives >= 2 && lives <= 3) {
        board[i][j] = 0b11; // Make the 2nd bit 1: 01 -> 11
      }
      if (board[i][j] === 0 && lives === 3) {
        board[i][j] = 0b10; // Make the 2nd bit 1: 00 -> 10
      }
    }
  }

  for (let i = 0; i < h; i++) {
    for (let j = 0; j < w; j++) {
      board[i][j] >>= 1; // Get the 2nd bit
    }
  }
};

File: Python/0289. Game of Life.py
# According to Wikipedia's article: "The Game of Life, also known simply as Life, is a cellular automaton devised by the British mathematician John Horton Conway in 1970."
# The board is made up of an m x n grid of cells, where each cell has an initial state: live (represented by a 1) or dead (represented by a 0). Each cell interacts with its eight neighbors (horizontal, vertical, diagonal) using the following four rules (taken from the above Wikipedia article):
#
# - Any live cell with fewer than two live neighbors dies as if caused by under-population.
# - Any live cell with two or three live neighbors lives on to the next generation.
# - Any live cell with more than three live neighbors dies, as if by over-population.
# - Any dead cell with exactly three live neighbors becomes a live cell, as if by reproduction.
#
# The next state is created by applying the above rules simultaneously to every cell in the current state, where births and deaths occur simultaneously. Given the current state of the m x n grid board, return the next state.
#
# Example 1:
#
# Input: board = [[0,1,0],[0,0,1],[1,1,1],[0,0,0]]
# Output: [[0,0,0],[1,0,1],[0,1,1],[0,1,0]]
#
# Example 2:
#
# Input: board = [[1,1],[1,0]]
# Output: [[1,1],[1,1]]
#
# Constraints:
#
# m == board.length
# n == board[i].length
# 1 <= m, n <= 25
# board[i][j] is 0 or 1.


# https://leetcode.com/problems/game-of-life/discuss/696582/Python-Very-Simple-Solution-with-Explanation-No-space-used
#  When the value needs to be updated, we don't just change  0 to 1 / 1 to 0 but we do in increments and decrements of 2. (table explains)
#  prev value  state change  curr state   curr val
#  0           no change     dead          0
#  1           no change     live          1
#  0           changed (+2)  live          2
#  1           changed (-2)  dead         -1
# Time O(m * n)
# Space O(1)
class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        m, n = len(board), len(board[0])
        dirs = [
            (-1, -1),
            (-1, 0),
            (-1, 1),
            (0, -1),
            (0, 1),
            (1, -1),
            (1, 0),
            (1, 1),
        ]

        for i in range(m):
            for j in range(n):
                live = 0  # live neighbors count
                for dx, dy in dirs:  # check and count neighbors in all directions
                    if (
                        0 <= i + dx < m
                        and 0 <= j + dy < n
                        and abs(board[i + dx][j + dy]) == 1
                    ):
                        live += 1
                if board[i][j] == 1 and (live < 2 or live > 3):  # Rule 1 or Rule 3
                    board[i][j] = -1
                if board[i][j] == 0 and live == 3:  # Rule 4
                    board[i][j] = 2
        for i in range(m):
            for j in range(n):
                board[i][j] = 1 if (board[i][j] > 0) else 0
        return board
